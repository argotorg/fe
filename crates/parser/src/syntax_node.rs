use crate::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FeLang {}

impl rowan::Language for FeLang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        SyntaxKind::try_from(raw.0).expect("invalid SyntaxKind from raw u16")
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<FeLang>;
pub type SyntaxToken = rowan::SyntaxToken<FeLang>;
pub type GreenNode = rowan::GreenNode;
pub type TextRange = rowan::TextRange;
pub type NodeOrToken = rowan::NodeOrToken<SyntaxNode, SyntaxToken>;
