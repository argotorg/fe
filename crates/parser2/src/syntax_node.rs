use crate::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FeLang {}

impl rowan::Language for FeLang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<FeLang>;
