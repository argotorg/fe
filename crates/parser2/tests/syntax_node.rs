use fe_parser2::syntax_node::SyntaxNode;

#[allow(unused)]
fn build_cst(input: &str) -> SyntaxNode {
    let (cst, errors) = fe_parser2::parse_source_file(input);
    for error in &errors {
        println!("{}", error.msg);
    }
    assert! {errors.is_empty()}
    cst
}

fe_compiler_test_utils::build_debug_snap_tests! {
    "parser2/test_files/syntax_node",
    "parser2/test_files/syntax_node",
    build_cst
}
