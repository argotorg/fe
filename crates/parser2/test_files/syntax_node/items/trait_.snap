---
source: crates/parser2/tests/syntax_node.rs
expression: snapshot
---
Root@0..538
  ItemList@0..538
    Trait@0..15
      TraitKw@0..5 "trait"
      WhiteSpace@5..6 " "
      Ident@6..12 "Marker"
      WhiteSpace@12..13 " "
      TraitItemList@13..15
        LBrace@13..14 "{"
        RBrace@14..15 "}"
    Newline@15..17 "\n\n"
    Trait@17..182
      ItemModifier@17..20
        PubKw@17..20 "pub"
      WhiteSpace@20..21 " "
      TraitKw@21..26 "trait"
      WhiteSpace@26..27 " "
      Ident@27..30 "Foo"
      WhiteSpace@30..31 " "
      TraitItemList@31..182
        LBrace@31..32 "{"
        Newline@32..33 "\n"
        WhiteSpace@33..37 "    "
        Fn@37..79
          FnKw@37..39 "fn"
          WhiteSpace@39..40 " "
          Ident@40..43 "foo"
          GenericParamList@43..67
            Lt@43..44 "<"
            GenericParam@44..52
              Ident@44..45 "T"
              TraitBoundList@45..52
                Colon@45..46 ":"
                WhiteSpace@46..47 " "
                TraitBound@47..52
                  Path@47..52
                    PathSegment@47..52
                      Ident@47..52 "Trait"
            Comma@52..53 ","
            WhiteSpace@53..54 " "
            GenericParam@54..66
              ConstKw@54..59 "const"
              WhiteSpace@59..60 " "
              Ident@60..61 "U"
              TraitBoundList@61..66
                Colon@61..62 ":"
                WhiteSpace@62..63 " "
                TraitBound@63..66
                  Path@63..66
                    PathSegment@63..66
                      Ident@63..66 "i32"
            Gt@66..67 ">"
          FnArgList@67..79
            LParen@67..68 "("
            FnArg@68..72
              Ident@68..69 "t"
              Colon@69..70 ":"
              WhiteSpace@70..71 " "
              PathType@71..72
                Path@71..72
                  PathSegment@71..72
                    Ident@71..72 "T"
            Comma@72..73 ","
            WhiteSpace@73..74 " "
            FnArg@74..78
              Ident@74..75 "u"
              Colon@75..76 ":"
              WhiteSpace@76..77 " "
              PathType@77..78
                Path@77..78
                  PathSegment@77..78
                    Ident@77..78 "U"
            RParen@78..79 ")"
        Newline@79..81 "\n\n"
        WhiteSpace@81..85 "    "
        Fn@85..180
          FnKw@85..87 "fn"
          WhiteSpace@87..88 " "
          Ident@88..102 "default_method"
          GenericParamList@102..116
            Lt@102..103 "<"
            GenericParam@103..115
              Ident@103..104 "T"
              TraitBoundList@104..115
                Colon@104..105 ":"
                WhiteSpace@105..106 " "
                TraitBound@106..109
                  Path@106..109
                    PathSegment@106..109
                      Ident@106..109 "Add"
                WhiteSpace@109..110 " "
                Plus@110..111 "+"
                WhiteSpace@111..112 " "
                TraitBound@112..115
                  Path@112..115
                    PathSegment@112..115
                      Ident@112..115 "Sub"
            Gt@115..116 ">"
          FnArgList@116..132
            LParen@116..117 "("
            FnArg@117..123
              Ident@117..120 "lhs"
              Colon@120..121 ":"
              WhiteSpace@121..122 " "
              PathType@122..123
                Path@122..123
                  PathSegment@122..123
                    Ident@122..123 "T"
            Comma@123..124 ","
            WhiteSpace@124..125 " "
            FnArg@125..131
              Ident@125..128 "rhs"
              Colon@128..129 ":"
              WhiteSpace@129..130 " "
              PathType@130..131
                Path@130..131
                  PathSegment@130..131
                    Ident@130..131 "T"
            RParen@131..132 ")"
          WhiteSpace@132..134 "  "
          Arrow@134..136 "->"
          WhiteSpace@136..137 " "
          PathType@137..140
            Path@137..140
              PathSegment@137..140
                Ident@137..140 "i32"
          WhiteSpace@140..141 " "
          BlockExpr@141..180
            LBrace@141..142 "{"
            Newline@142..143 "\n"
            WhiteSpace@143..151 "        "
            ExprStmt@151..174
              BinExpr@151..174
                BinExpr@151..160
                  Path@151..154
                    PathSegment@151..154
                      Ident@151..154 "lhs"
                  WhiteSpace@154..155 " "
                  Plus@155..156 "+"
                  WhiteSpace@156..157 " "
                  Path@157..160
                    PathSegment@157..160
                      Ident@157..160 "lhs"
                WhiteSpace@160..161 " "
                Minus@161..162 "-"
                WhiteSpace@162..163 " "
                ParenExpr@163..174
                  LParen@163..164 "("
                  BinExpr@164..173
                    Path@164..167
                      PathSegment@164..167
                        Ident@164..167 "rhs"
                    WhiteSpace@167..168 " "
                    Plus@168..169 "+"
                    WhiteSpace@169..170 " "
                    Path@170..173
                      PathSegment@170..173
                        Ident@170..173 "rhs"
                  RParen@173..174 ")"
            Newline@174..175 "\n"
            WhiteSpace@175..179 "    "
            RBrace@179..180 "}"
        Newline@180..181 "\n"
        RBrace@181..182 "}"
    Newline@182..184 "\n\n"
    Trait@184..241
      ItemModifier@184..187
        PubKw@184..187 "pub"
      WhiteSpace@187..188 " "
      TraitKw@188..193 "trait"
      WhiteSpace@193..194 " "
      Ident@194..197 "Add"
      GenericParamList@197..202
        Lt@197..198 "<"
        GenericParam@198..201
          Ident@198..201 "RHS"
        Gt@201..202 ">"
      WhiteSpace@202..203 " "
      TraitItemList@203..241
        LBrace@203..204 "{"
        Newline@204..205 "\n"
        WhiteSpace@205..209 "    "
        Fn@209..239
          FnKw@209..211 "fn"
          WhiteSpace@211..212 " "
          Ident@212..215 "add"
          FnArgList@215..231
            LParen@215..216 "("
            FnArg@216..220
              SelfKw@216..220 "self"
            Comma@220..221 ","
            WhiteSpace@221..222 " "
            FnArg@222..230
              Ident@222..225 "rhs"
              Colon@225..226 ":"
              WhiteSpace@226..227 " "
              PathType@227..230
                Path@227..230
                  PathSegment@227..230
                    Ident@227..230 "Rhs"
            RParen@230..231 ")"
          WhiteSpace@231..232 " "
          Arrow@232..234 "->"
          WhiteSpace@234..235 " "
          SelfType@235..239
            SelfTypeKw@235..239 "Self"
        Newline@239..240 "\n"
        RBrace@240..241 "}"
    Newline@241..244 "\n\n\n"
    Trait@244..325
      ItemModifier@244..247
        PubKw@244..247 "pub"
      WhiteSpace@247..248 " "
      TraitKw@248..253 "trait"
      WhiteSpace@253..254 " "
      Ident@254..259 "Parse"
      WhiteSpace@259..260 " "
      TraitItemList@260..325
        LBrace@260..261 "{"
        Newline@261..262 "\n"
        WhiteSpace@262..266 "    "
        Fn@266..323
          FnKw@266..268 "fn"
          WhiteSpace@268..269 " "
          Ident@269..274 "parse"
          GenericParamList@274..290
            Lt@274..275 "<"
            GenericParam@275..289
              Ident@275..276 "S"
              TraitBoundList@276..289
                Colon@276..277 ":"
                WhiteSpace@277..278 " "
                TraitBound@278..289
                  Path@278..289
                    PathSegment@278..289
                      Ident@278..289 "TokenStream"
            Gt@289..290 ">"
          FnArgList@290..323
            LParen@290..291 "("
            FnArg@291..299
              MutKw@291..294 "mut"
              WhiteSpace@294..295 " "
              SelfKw@295..299 "self"
            Comma@299..300 ","
            WhiteSpace@300..301 " "
            FnArg@301..322
              MutKw@301..304 "mut"
              WhiteSpace@304..305 " "
              Ident@305..311 "parser"
              Colon@311..312 ":"
              WhiteSpace@312..313 " "
              PathType@313..322
                Path@313..319
                  PathSegment@313..319
                    Ident@313..319 "Parser"
                GenericArgList@319..322
                  Lt@319..320 "<"
                  GenericArg@320..321
                    PathType@320..321
                      Path@320..321
                        PathSegment@320..321
                          Ident@320..321 "S"
                  Gt@321..322 ">"
            RParen@322..323 ")"
        Newline@323..324 "\n"
        RBrace@324..325 "}"
    Newline@325..327 "\n\n"
    Impl@327..538
      ImplKw@327..331 "impl"
      WhiteSpace@331..332 " "
      PathType@332..354
        Path@332..338
          PathSegment@332..338
            Ident@332..338 "Parser"
        GenericArgList@338..354
          Lt@338..339 "<"
          GenericArg@339..353
            PathType@339..340
              Path@339..340
                PathSegment@339..340
                  Ident@339..340 "S"
            TraitBoundList@340..353
              Colon@340..341 ":"
              WhiteSpace@341..342 " "
              TraitBound@342..353
                Path@342..353
                  PathSegment@342..353
                    Ident@342..353 "TokenStream"
          Gt@353..354 ">"
      WhiteSpace@354..355 " "
      ImplItemList@355..538
        LBrace@355..356 "{"
        Newline@356..357 "\n"
        WhiteSpace@357..361 "    "
        Fn@361..536
          ItemModifier@361..364
            PubKw@361..364 "pub"
          WhiteSpace@364..365 " "
          FnKw@365..367 "fn"
          WhiteSpace@367..368 " "
          Ident@368..373 "parse"
          GenericParamList@373..383
            Lt@373..374 "<"
            GenericParam@374..382
              Ident@374..375 "T"
              TraitBoundList@375..382
                Colon@375..376 ":"
                WhiteSpace@376..377 " "
                TraitBound@377..382
                  Path@377..382
                    PathSegment@377..382
                      Ident@377..382 "Parse"
            Gt@382..383 ">"
          FnArgList@383..439
            LParen@383..384 "("
            FnArg@384..392
              MutKw@384..387 "mut"
              WhiteSpace@387..388 " "
              SelfKw@388..392 "self"
            Comma@392..393 ","
            WhiteSpace@393..394 " "
            FnArg@394..406
              MutKw@394..397 "mut"
              WhiteSpace@397..398 " "
              Ident@398..403 "scope"
              Colon@403..404 ":"
              WhiteSpace@404..405 " "
              PathType@405..406
                Path@405..406
                  PathSegment@405..406
                    Ident@405..406 "T"
            Comma@406..407 ","
            WhiteSpace@407..408 " "
            FnArg@408..438
              Ident@408..418 "checkpoint"
              Colon@418..419 ":"
              WhiteSpace@419..420 " "
              PathType@420..438
                Path@420..426
                  PathSegment@420..426
                    Ident@420..426 "Option"
                GenericArgList@426..438
                  Lt@426..427 "<"
                  GenericArg@427..437
                    PathType@427..437
                      Path@427..437
                        PathSegment@427..437
                          Ident@427..437 "Checkpoint"
                  Gt@437..438 ">"
            RParen@438..439 ")"
          WhiteSpace@439..440 " "
          Arrow@440..442 "->"
          WhiteSpace@442..443 " "
          TupleType@443..461
            LParen@443..444 "("
            PathType@444..448
              Path@444..448
                PathSegment@444..448
                  Ident@444..448 "bool"
            Comma@448..449 ","
            WhiteSpace@449..450 " "
            PathType@450..460
              Path@450..460
                PathSegment@450..460
                  Ident@450..460 "Checkpoint"
            RParen@460..461 ")"
          WhiteSpace@461..462 " "
          BlockExpr@462..536
            LBrace@462..463 "{"
            Newline@463..464 "\n"
            WhiteSpace@464..472 "        "
            ExprStmt@472..530
              TupleExpr@472..530
                LParen@472..473 "("
                CallExpr@473..516
                  Path@473..493
                    PathSegment@473..483
                      Ident@473..483 "SyntaxNode"
                    Colon2@483..485 "::"
                    PathSegment@485..493
                      Ident@485..493 "new_root"
                  CallArgList@493..516
                    LParen@493..494 "("
                    CallArg@494..515
                      MethodCallExpr@494..515
                        FieldExpr@494..506
                          Path@494..498
                            PathSegment@494..498
                              SelfKw@494..498 "self"
                          Dot@498..499 "."
                          Ident@499..506 "builder"
                        Dot@506..507 "."
                        Ident@507..513 "finish"
                        CallArgList@513..515
                          LParen@513..514 "("
                          RParen@514..515 ")"
                    RParen@515..516 ")"
                Comma@516..517 ","
                WhiteSpace@517..518 " "
                FieldExpr@518..529
                  Path@518..522
                    PathSegment@518..522
                      SelfKw@518..522 "self"
                  Dot@522..523 "."
                  Ident@523..529 "errors"
                RParen@529..530 ")"
            Newline@530..531 "\n"
            WhiteSpace@531..535 "    "
            RBrace@535..536 "}"
        Newline@536..537 "\n"
        RBrace@537..538 "}"

